application/src/main/java/org/opentripplanner/graph_builder/module/osm/OsmAreaGroup.java

```diff
@@ -175,31 +196,27 @@ private static void processRing(
     Ring ring,
     Multimap<OsmNode, OsmWay> barriers,
     Multimap<OsmNodePair, OsmArea> areasForNodePair,
-    Map<OsmArea, Map<OsmWay, Set<OsmNode>>> barriersForArea
+    Map<OsmArea, Map<OsmWay, Set<OsmNodePair>>> barriersForArea
   ) {
     var nodes = ring.nodes;
     // the end node of a ring must be the same of the start node
     for (var i = 0; i < nodes.size() - 1; i++) {
-      var pair = new OsmNodePair(nodes.get(i), nodes.get(i + 1));
+      OsmNode node = nodes.get(i);
+      OsmNode nextNode = nodes.get(i + 1);
+      var pair = new OsmNodePair(node, nextNode);
       areasForNodePair.put(pair, area);
-    }
-    for (OsmNode node : ring.nodes) {
-      for (OsmNode anotherNode : ring.nodes) {
-        if (anotherNode != node) {
-          var sharedBarriers = getBarriers(barriers, node, anotherNode);
-          for (var barrier : sharedBarriers) {
-            if (!barriersForArea.containsKey(area)) {
-              barriersForArea.put(area, new HashMap<>());
-            }
-            var barrierMap = barriersForArea.get(area);
-            if (!barrierMap.containsKey(barrier)) {
-              barrierMap.put(barrier, new HashSet<>());
-            }
-            var nodesOnBarrier = barrierMap.get(barrier);
-            nodesOnBarrier.add(node);
-            nodesOnBarrier.add(anotherNode);
-          }
+
+      var sharedBarriers = getBarriers(barriers, node, nextNode);
+      for (var barrier : sharedBarriers) {
+        if (!barriersForArea.containsKey(area)) {
+          barriersForArea.put(area, new HashMap<>());
+        }
+        var barrierMap = barriersForArea.get(area);
+        if (!barrierMap.containsKey(barrier)) {
+          barrierMap.put(barrier, new HashSet<>());
         }
+        var nodesOnBarrier = barrierMap.get(barrier);
+        nodesOnBarrier.add(new OsmNodePair(node, nextNode));
       }
```

Comment:
This is really a minor issue, but can't we use already constructed node pair object `pair `here as well?